File: backend/src/DVC.Domain/Entities/Assignment.cs
using System;

namespace DVC.Domain.Entities
{
    public enum AssignmentStatus
    {
        Assigned,
        Dispatched,
        InProgress,
        Completed,
        Cancelled
    }

    public class Assignment
    {
        public Guid Id { get; set; } = Guid.NewGuid();

        // Incident associated with this assignment
        public Guid IncidentId { get; set; }
        public Incident? Incident { get; set; }

        // Volunteer assigned to the incident
        public Guid VolunteerId { get; set; }
        public User? Volunteer { get; set; }

        // Matching proposal that resulted in this assignment
        public Guid MatchId { get; set; }
        public VolunteerMatch? Match { get; set; }

        public AssignmentStatus Status { get; set; } = AssignmentStatus.Assigned;

        // Used by the time-window validation
        public int EstimatedDurationMinutes { get; set; }

        public DateTime AssignedAtUtc { get; set; } = DateTime.UtcNow;
        public DateTime? DispatchedAtUtc { get; set; }
        public DateTime? StartedAtUtc { get; set; }
        public DateTime? CompletedAtUtc { get; set; }

        public DateTime CreatedAtUtc { get; set; } = DateTime.UtcNow;
    }
}

File: mobile/lib/models/assignment.dart
enum AssignmentStatus {
  assigned,
  dispatched,
  inProgress,
  completed,
  cancelled,
}

class Assignment {
  final String id;
  final String incidentId;
  final String volunteerId;
  final String matchId;
  final AssignmentStatus status;
  final int estimatedDurationMinutes;
  final DateTime assignedAtUtc;
  final DateTime? dispatchedAtUtc;
  final DateTime? startedAtUtc;
  final DateTime? completedAtUtc;
  final DateTime createdAtUtc;

  Assignment({
    required this.id,
    required this.incidentId,
    required this.volunteerId,
    required this.matchId,
    required this.status,
    required this.estimatedDurationMinutes,
    required this.assignedAtUtc,
    this.dispatchedAtUtc,
    this.startedAtUtc,
    this.completedAtUtc,
    required this.createdAtUtc,
  });

  factory Assignment.fromJson(Map<String, dynamic> json) {
    return Assignment(
      id: json['id'],
      incidentId: json['incidentId'],
      volunteerId: json['volunteerId'],
      matchId: json['matchId'],
      status: _parseStatus(json['status']),
      estimatedDurationMinutes: json['estimatedDurationMinutes'],
      assignedAtUtc: DateTime.parse(json['assignedAtUtc']),
      dispatchedAtUtc: json['dispatchedAtUtc'] != null
          ? DateTime.parse(json['dispatchedAtUtc'])
          : null,
      startedAtUtc: json['startedAtUtc'] != null
          ? DateTime.parse(json['startedAtUtc'])
          : null,
      completedAtUtc: json['completedAtUtc'] != null
          ? DateTime.parse(json['completedAtUtc'])
          : null,
      createdAtUtc: DateTime.parse(json['createdAtUtc']),
    );
  }

  static AssignmentStatus _parseStatus(dynamic value) {
    switch (value.toString().toLowerCase()) {
      case 'assigned':
        return AssignmentStatus.assigned;
      case 'dispatched':
        return AssignmentStatus.dispatched;
      case 'inprogress':
      case 'in_progress':
        return AssignmentStatus.inProgress;
      case 'completed':
        return AssignmentStatus.completed;
      case 'cancelled':
        return AssignmentStatus.cancelled;
      default:
        throw FormatException('Unknown assignment status: $value');
    }
  }

  String get statusText {
    switch (status) {
      case AssignmentStatus.assigned:
        return 'Assigned';
      case AssignmentStatus.dispatched:
        return 'Dispatched';
      case AssignmentStatus.inProgress:
        return 'In Progress';
      case AssignmentStatus.completed:
        return 'Completed';
      case AssignmentStatus.cancelled:
        return 'Cancelled';
    }
  }
}